src/main/java/io/confluent/connect/elasticsearch/OffsetTracker.java

```diff
@@ -0,0 +1,180 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.connect.elasticsearch;
+
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.connect.errors.ConnectException;
+import org.apache.kafka.connect.sink.ErrantRecordReporter;
+import org.apache.kafka.connect.sink.SinkRecord;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Supplier;
+
+import static java.util.stream.Collectors.toMap;
+
+class OffsetTracker {
+
+  private static final Logger log = LoggerFactory.getLogger(OffsetTracker.class);
+
+  static class OffsetState {
+
+    private static final Supplier<Boolean> TRUE = () -> true;
+    private final long offset;
+    private final AtomicReference<Supplier<Boolean>> processed = new AtomicReference<>();
+
+    OffsetState(long offset) {
+      this.offset = offset;
+    }
+
+    /**
+     * Marks the offset as processed (ready to report to preCommit), if it has not been marked yet
+     */
+    public void markProcessed() {
+      markProcessed(TRUE);
+    }
+
+    /**
+     * Provides a supplier that will determine if the record was processed or not.
+     * It does nothing is markProcessed was called before.
+     * Intended for async operations that return a Future and don't provide a callback we can
+     * hook to. (e.g. {@link ErrantRecordReporter#report(SinkRecord, Throwable)})
+     */
+    public void markProcessed(Supplier<Boolean> processedSupplier) {
+      processed.compareAndSet(null, processedSupplier);
+    }
+
+    public boolean isProcessed() {
+      Supplier<Boolean> booleanSupplier = processed.get();
+      return booleanSupplier != null && Boolean.TRUE.equals(booleanSupplier.get());
+    }
+  }
+
+  private final Map<TopicPartition, Map<Long, OffsetState>> offsetsByPartition
+          = new ConcurrentHashMap<>();
+
+  private final Map<TopicPartition, Long> maxOffsetByPartition
+          = new ConcurrentHashMap<>();
+
+  private final AtomicLong numEntries = new AtomicLong();
+
+  private static final long LIMIT = 1_000; // TODO calculate / config
+
+  // fairness is not important, things can only be blocked from one thread
+  private final ReentrantLock lock = new ReentrantLock();
+  private final Condition notFull = lock.newCondition();
```

Comment:
Any reason not to use the intrinsic lock and wait/notify API? Seems like you are locking the entire methods anyways, might be more readable to just have them as synchronized methods.